spec/utils/db_poller_spec.rb

```diff
@@ -0,0 +1,233 @@
+# frozen_string_literal: true
+
+each_db_config(Deimos::Utils::DbPoller) do
+  describe '#start!' do
+
+    before(:each) do
+      producer_class = Class.new(Deimos::Producer) do
+        schema 'MySchema'
+        namespace 'com.my-namespace'
+        topic 'my-topic'
+        key_config field: 'test_id'
+      end
+      stub_const('MyProducer', producer_class)
+
+      producer_class = Class.new(Deimos::Producer) do
+        schema 'MySchemaWithId'
+        namespace 'com.my-namespace'
+        topic 'my-topic'
+        key_config plain: true
+      end
+      stub_const('MyProducerWithID', producer_class)
+    end
+
+    it 'should raise an error if no pollers configured' do
+      Deimos.configure {}
+      expect { described_class.start! }.to raise_error('No pollers configured!')
+    end
+
+    it 'should start pollers as configured' do
+      Deimos.configure do
+        db_poller do
+          producer_class 'MyProducer'
+        end
+        db_poller do
+          producer_class 'MyProducerWithID'
+        end
+      end
+
+      allow(Deimos::Utils::DbPoller).to receive(:new)
+      signal_double = instance_double(Deimos::Utils::SignalHandler, run!: nil)
+      allow(Deimos::Utils::SignalHandler).to receive(:new).and_return(signal_double)
+      described_class.start!
+      expect(Deimos::Utils::DbPoller).to have_received(:new).twice
+      expect(Deimos::Utils::DbPoller).to have_received(:new).
+        with(Deimos.config.db_poller_objects[0])
+      expect(Deimos::Utils::DbPoller).to have_received(:new).
+        with(Deimos.config.db_poller_objects[1])
+    end
+  end
+
+  describe 'pollers' do
+    include_context 'with widgets'
+
+    let(:poller) do
+      poller = described_class.new(config)
+      allow(poller).to receive(:sleep)
+      poller
+    end
+
+    let(:config) { Deimos.config.db_poller_objects.first }
+
+    before(:each) do
+      producer_class = Class.new(Deimos::ActiveRecordProducer) do
+        schema 'MySchemaWithId'
+        namespace 'com.my-namespace'
+        topic 'my-topic-with-id'
+        key_config none: true
+        record_class Widget
+
+        # :nodoc:
+        def self.generate_payload(attrs, widget)
+          super.merge(message_id: widget.generated_id)
+        end
+      end
+      stub_const('MyProducer', producer_class)
+
+      Deimos.configure do
+        db_poller do
+          producer_class 'MyProducer'
+          run_every 1.minute
+        end
+      end
+    end
+
+    after(:each) do
+      Deimos::PollInfo.delete_all
+      Widget.delete_all
+      travel_back
+    end
+
+    it 'should crash if initialized with an invalid producer' do
+      config.producer_class = 'NoProducer'
+      expect { described_class.new(config) }.to raise_error('Class NoProducer not found!')
+    end
+
+    specify '#retrieve_poll_info' do
+      travel_to(Time.zone.local(2020, 0o5, 0o5, 0o5, 0o0, 0o0))
```

Comment:
Replace all the ugly octals in this file (lots below too)

Reply:
lol missed these. Thanks Rubocop